python/ql/src/semmle/python/frameworks/Peewee.qll

```diff
@@ -0,0 +1,192 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `peewee` PyPI package.
+ * See
+ * - https://pypi.org/project/peewee/
+ * - https://docs.peewee-orm.com/en/latest/index.html
+ */
+
+private import python
+private import semmle.python.dataflow.new.DataFlow
+private import semmle.python.dataflow.new.TaintTracking
+private import semmle.python.Concepts
+private import semmle.python.ApiGraphs
+private import semmle.python.frameworks.PEP249
+
+/**
+ * Provides models for the `peewee` PyPI package.
+ * See
+ * - https://pypi.org/project/peewee/
+ * - https://docs.peewee-orm.com/en/latest/index.html
+ */
+private module Peewee {
+  /** Provides models for the `peewee.Database` class and subclasses. */
+  module Database {
+    /** Gets a reference to the `peewee.Database` class or any subclass. */
+    API::Node subclassRef() {
+      result = API::moduleImport("peewee").getMember("Database").getASubclass*()
+      or
+      // known subclasses
+      result =
+        API::moduleImport("peewee")
+            .getMember(["SqliteDatabase", "MySQLDatabase", "PostgresqlDatabase"])
+            .getASubclass*()
+      or
+      // Ohter known subclasses, semi auto generated by using
+      // ```codeql
+      // class DBClass extends Class, SelfRefMixin {
+      //   DBClass() {
+      //     exists(this.getLocation().getFile().getRelativePath()) and
+      //     this.getName().matches("%Database") and
+      //     this.getABase().(Name).getId().matches("%Database")
+      //   }
+      // }
+      //
+      // from DBClass dbClass, Module mod
+      // where
+      //   dbClass.getScope() = mod
+      // select mod.getName()+ "." + dbClass.getName()
+      // ```
+      result =
+        API::moduleImport("playhouse")
+            .getMember("apsw_ext")
+            .getMember("APSWDatabase")
+            .getASubclass*()
+      or
+      result =
+        API::moduleImport("playhouse")
+            .getMember("cockroachdb")
+            .getMember("CockroachDatabase")
+            .getASubclass*()
+      or
+      result =
+        API::moduleImport("playhouse")
+            .getMember("cockroachdb")
+            .getMember("PooledCockroachDatabase")
+            .getASubclass*()
```

Comment:
This list is quite long. It could become about half the size with rewrites like the following:
```suggestion
      result =
        API::moduleImport("playhouse")
            .getMember("cockroachdb")
            .getMember(["CockroachDatabase", "PooledCockroachDatabase"])
            .getASubclass*()
```

Reply:
As highlighted by the comment above, these are _semi auto generated_, so I don't think it's worth it to go through this trouble.